### ISSUE TITLE
BUG - Bad Reconstruction When Loading Projections from .tiff_stack

### ISSUE THREAD

**Issue by WilliamJudge94** [tomopy/tomopy#525]
**Describe the problem**
When a User tries to reconstruct a single slice tomopy works great. When a User tries to loop through more than one slice (necessary for obtaining a full reconstruction) tomopy will not give consistent results when compared to slices run individually. 

**To Reproduce (if applicable)**

https://colab.research.google.com/drive/1ALFxIlzjaUzn1YZtTbERVg81kzh6ePlY?usp=sharing

**Expected behavior (if applicable)**
For a User to be able to loop through all slices of a tomographic projection and have the results be identical to those run outside of a loop.

**Helpful images (if applicable)**

Both images are tomopy reconstructions of identical slices of the raw projection data. This first was run through an iterative loop, which should run through all available projection slices:

<img width="432" alt="loaded_through_loop" src="https://user-images.githubusercontent.com/37912237/97046641-4a612180-153d-11eb-9261-6ea3f49e4746.png">

The second image was the identical slice data, passed into tomopy, but not inside of a loop. These should produce the same result, but they are wildly different. 

<img width="432" alt="one_off_load" src="https://user-images.githubusercontent.com/37912237/97046650-4d5c1200-153d-11eb-85eb-87626783c41e.png">

**Platform Information:**
 - OS:  Ubuntu
 - Python Version e.g. 3.6.8
 - TomoPy Version 1.2.1

**Additional context**

**Comment by carterbox** [comment]
Hi @WilliamJudge94,

I be happy to help with this problem.

1. Have you checked that the arrays loaded from the `tiff` and the `hdf` files are exactly the same? You can use `numpy.array_equal()`.

1. Could you provide an actual `tiff` file for which this problem manifests or update your Collaboratory notebook environment so it has all the necessary packages to run?

**Comment by WilliamJudge94** [comment]
I have just added you to the google drive file which houses the raw data. I have also updated the Google Collab document on how I installed all dependencies. Unfortunately, Collab is being a little buggy and won't import tomopy, so you might have to use it as a template, and run the corresponding code on your own machine. 

To answer your questions:

1. the .tif and the raw hdf data is not the same. I load in the hdf data, pre-process, then continue to use tomopy. For the tiff files these were already reprocessed so there is no need to pre-process again. 

Although the main issue described in this post is when I use the SAME .tiff stack, the reconstruction works well when I do a single reconstruction by itself. Example:

`row = 5 
recon = tomopy.recon(full_data[:, (row, row+1)])`

But as soon as I put the code into a loop (to loop through all the different rows), then a lot of the reconstructed images are of very poor quality. Example:

`for row in range(10):
    recon = tomopy.recon(full_data[:, (row, row+1)])`

when row = 5 in both cases, they might equal each other OR the image in the loop will be blank, random edge circle, blurry, etc. 

2. Sent through a shared drive

3. astra == 1.8

**Comment by carterbox** [comment]
> Although the main issue described in this post is when I use the SAME .tiff stack

Why tell me about the files loading process if the problem has nothing to do with file loading? Can you reproduce the problem with simulated data?

**Comment by WilliamJudge94** [comment]
Sorry, I updated the original post to be more clear. I hope this helps clarify the issue.

**Comment by carterbox** [comment]
```python

# There are 1210 projections inside the hdf5 file
prj_data = dxchange.read_tiff('/path/to/raw_data.tif')

# Making angle data
theta = tomopy.angles(prj_data.shape[0], 0, 180)


final = 50
#final = 1024 

for row in tqdm(range(0, final)):

    slc_proj = reconstruct_single_slice(prj_data.copy(), 
                                        theta,
                                        rot_center = 600, 
                                        rows=(row, row+1), )
    
    
    for idx, im in enumerate(slc_proj):
        fig = plt.figure(figsize=(10, 10))
        plt.imshow(im)
        plt.show()
```
Your function `reconstruct_single_slice()` reconstructs 2 slices, so you are reconstructing many slices twice. For example, in the code chunk above, you would reconstruct rows 0, 1; then rows 1, 2; then rows 2, 3... Is that the problem?

**Comment by carterbox** [comment]
For example: `full_data[:, (row, row+1)]` is not equivalent to `full_data[:, row:row+1]`. The first is a list of indices, the second is a range where the end is not included.

**Comment by WilliamJudge94** [comment]
Changing the loop to say `for row in np.arange(0, num, 2)` where `num=number_of_slices_to_reconstruct` helps the problem, but does not eliminate it. There are still reconstruction slices that "error out" with features resembling those found in the attached image:

<img width="275" alt="Screen Shot 2020-10-23 at 6 16 34 PM" src="https://user-images.githubusercontent.com/37912237/97061285-fa458780-155b-11eb-9c0d-7efde8faba0b.png">

**Comment by WilliamJudge94** [comment]
I forgot to mention another aspect of this bug. During the reconstruction, it will frequently "Kernel Restart" without raising any other errors. The only warning I receive during the reconstruction is shown below. No OOM for the RAM or VRAM, no other python flags to let me know what is going wrong with the reconstruction. Just the Kernel Restart popup you see below. 


<img width="566" alt="Screen Shot 2020-10-24 at 2 36 06 PM" src="https://user-images.githubusercontent.com/37912237/97092088-8613ee00-1606-11eb-81f3-586cfee651aa.png">

**Comment by carterbox** [comment]
Use `ncore=1`, and check that you aren't running out of GPU because you are using GPU methods through the astra-toolbox.

**Comment by WilliamJudge94** [comment]
If it was an OOM I would assume

 1) to see an OOM error (which I do not) 

2) The "kernel restarting" error to happen on the same rough iteration step, but sometimes the reconstruction lasts 20 seconds before hitting the "kernel restarting" error, and other times it will last an hour. (I tested this error on both an RTX 2080ti (11gb VRAM) and an RTX quadro 8000 (48gb VRAM). There "kernel restarting" error persists on both GPU's)

However all errors seem to stop when I switch to the gridrec algorithm so I am assuming there is an issue with astra

**Comment by carterbox** [comment]
Don't assume that mixed-language libraries handle errors "correctly" / in a Pythonic way. It's extra effort for a developer to catch and translate errors into Python exceptions.

Use `ncore=1`. Probably the astra-toolbox doesn't support being run on 16 simultaneous CPU threads.

**Comment by carterbox** [comment]
@WilliamJudge94, Is your problem solved?

**Comment by WilliamJudge94** [comment]
ncore=1 did not stop the error. Although switching to algorithm='gridrec' did

**Comment by carterbox** [comment]
If you run the following script, do you observe the error?

```python
import tomopy
import dxchange


def tomo_recon(prj, theta, rot_center):

    ##---------------------------------------ASTRA GPU based reconstruction--------------------------------------------------------------##
    ##---Only the SIRT, Simultaneous Iterative Reconstruction Technique, GPU based algorithm for tomographic reconstruction is seen here-##
    ##------------------see TomoPy documentation for more GPU and CPU based algorthim options--------------------------------------------##

    #GPU-Based SIRT method with nonnegativity constraint
    extra_options = {'MinConstraint': 0}

    #SIRT GPU algorthim, make sure to define the optimized iteration number
    options = {
        'proj_type': 'cuda',
        'method': 'SIRT_CUDA',
        'num_iter': 200,
        'extra_options': extra_options,
    }
    recon = tomopy.recon(
        prj,
        theta,
        center=rot_center,
        algorithm=tomopy.astra,
        ncore=1,
        options=options,
    )

    ###Optional Post Reconstruction Artifact Corrections###

    #Apply circle mask to reconstructed data
    recon = tomopy.circ_mask(recon, axis=0, ratio=0.95)

    #Remove ring artifacts, this comes with a slight resolution cost
    #recon = tomopy.remove_ring(recon, center_x=None, center_y=None, thresh=300.0)

    return recon


def reconstruct_single_slice(prj_data,
                             theta,
                             rows=(604, 606),
                             rot_center=True):

    prj = prj_data[:, rows]

    recon = tomo_recon(prj, theta, rot_center=rot_center)

    return recon


if __name__ == "__main__":

    # There are 1210 projections inside the hdf5 file
    prj_data = dxchange.read_tiff('raw_data.tif')

    # Making angle data
    theta = tomopy.angles(prj_data.shape[0], 0, 180)

    final = 1024

    for row in range(0, final):

        slc_proj = reconstruct_single_slice(
            prj_data,
            theta,
            rot_center=600,
            rows=(row,),
        )

        for idx, im in enumerate(slc_proj):
            dxchange.write_tiff(im, f'{row + idx:04d}')
```

**Comment by carterbox** [comment]
Also, please update to the latest versions of TomoPy=1.9 and ASTRA=1.9.

**Comment by WilliamJudge94** [comment]
Unfortunately, I have only been able to install astra 1.8 and tomopy 1.2 due to python compatibility issues (python3.6 vs 3.7 vs 3.8). There is an install error for one of the packages when I update to python3.8. So I am forced to use python3.6. Which, then conda only downloads/installs astra 1.8 and tomopy 1.2

**Comment by carterbox** [comment]
Please install `astra-toolbox=1.9` [from their dev channel](https://www.astra-toolbox.com/downloads/index.html). Because they haven't released a new build of `astra-toolbox=1.8` since 2017, they explicitly recommend version 1.9 (even though it is "unreleased").

```bash
conda install -c astra-toolbox/label/dev astra-toolbox
```
